File: litho-editor-core/src/main/java/com/facebook/litho/editor/SimpleEditor.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho.editor;

import com.facebook.litho.editor.instances.EditorUtils;
import com.facebook.litho.editor.model.EditorArray;
import com.facebook.litho.editor.model.EditorBool;
import com.facebook.litho.editor.model.EditorNumber;
import com.facebook.litho.editor.model.EditorPick;
import com.facebook.litho.editor.model.EditorShape;
import com.facebook.litho.editor.model.EditorString;
import com.facebook.litho.editor.model.EditorValue;
import java.lang.reflect.Field;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import javax.annotation.Nullable;

public final class SimpleEditor {

  private SimpleEditor() {}

  // Constructors

  public static <T> Editor makeMutable(final MutablePropertyEditor<T> propertyEditor) {
    return new Editor() {
      @Override
      public EditorValue read(Field f, Object node) {
        return makeValue(f, node, propertyEditor);
      }

      @Override
      public boolean write(final Field f, final Object node, final EditorValue values) {
        final T value = EditorUtils.getNodeUNSAFE(f, node);
        values.when(
            new EditorValue.DefaultEditorVisitor() {
              @Override
              public Void isShape(EditorShape object) {
                final EditorValue editorValue = object.value.get(f.getName());
                return editorValue != null
                    ? editorValue.when(writeMutable(propertyEditor, value))
                    : null;
              }
            });
        return true;
      }
    };
  }

  public static <T> Editor makeImmutable(final ImmutablePropertyEditor<T> propertyEditor) {
    return new Editor() {
      @Override
      public EditorValue read(Field f, Object node) {
        return makeValue(f, node, propertyEditor);
      }

      @Override
      public boolean write(final Field f, final Object node, final EditorValue values) {
        final T value = EditorUtils.getNodeUNSAFE(f, node);
        values.when(
            new EditorValue.DefaultEditorVisitor() {
              @Override
              public Void isShape(EditorShape object) {
                return object
                    .value
                    .get(f.getName())
                    .when(writeImmutable(f, node, propertyEditor, value));
              }
            });
        return true;
      }
    };
  }

  // Read data

  private static <T> EditorValue makeValue(Field f, Object node, PropertyReader<T> propertyReader) {
    T value = EditorUtils.getNodeUNSAFE(f, node);
    final Map<String, SimpleEditorValue> properties = propertyReader.readProperties(value);
    Map<String, EditorValue> shape = new HashMap<>();
    for (Map.Entry<String, SimpleEditorValue> property : properties.entrySet()) {
      shape.put(property.getKey(), property.getValue().value);
    }
    return EditorValue.shape(shape);
  }

  // Write data

  private static <T> EditorValue.DefaultEditorVisitor writeMutable(
      final MutablePropertyEditor<T> propertyEditor, final T value) {
    return new EditorValue.DefaultEditorVisitor() {
      @Override
      public Void isShape(EditorShape shape) {
        final Map<String, SimpleEditorValue> properties = propertyEditor.readProperties(value);
        final Map<String, EditorValue> updates = shape.value;
        for (final Map.Entry<String, EditorValue> updatedProperty : updates.entrySet()) {
          final SimpleEditorValue newValue =
              SimpleEditorValue.fromEditorValueOrNull(updatedProperty.getValue());
          final SimpleEditorValue oldValue = properties.get(updatedProperty.getKey());
          if (newValue != null && oldValue != null && newValue.type == oldValue.type) {
            newValue.value.whenPrimitive(
                new EditorValue.EditorPrimitiveVisitor() {
                  @Override
                  public boolean isNumber(String[] path, EditorNumber number) {
                    propertyEditor.writeNumberProperty(
                        value, updatedProperty.getKey(), number.value);
                    return true;
                  }

                  @Override
                  public boolean isString(String[] path, EditorString string) {
                    propertyEditor.writeStringProperty(
                        value, updatedProperty.getKey(), string.value);
                    return false;
                  }

                  @Override
                  public boolean isBool(String[] path, EditorBool bool) {
                    propertyEditor.writeBoolProperty(value, updatedProperty.getKey(), bool.value);
                    return false;
                  }

                  @Override
                  public boolean isPick(String[] path, EditorPick pick) {
                    propertyEditor.writePickProperty(
                        value, updatedProperty.getKey(), pick.selected);
                    return false;
                  }
                });
          }
        }
        return null;
      }
    };
  }

  private static <T> EditorValue.EditorVisitor<Void> writeImmutable(
      final Field f,
      final Object node,
      final ImmutablePropertyEditor<T> propertyEditor,
      final T value) {
    return new EditorValue.DefaultEditorVisitor() {
      @Override
      public Void isShape(EditorShape shape) {
        final Map<String, SimpleEditorValue> properties = propertyEditor.readProperties(value);
        final Map<String, String> stringProperties = new HashMap<>();
        final Map<String, Number> numberProperties = new HashMap<>();
        final Map<String, Boolean> boolProperties = new HashMap<>();
        final Map<String, String> pickProperties = new HashMap<>();
        for (final Map.Entry<String, SimpleEditorValue> property : properties.entrySet()) {
          property
              .getValue()
              .value
              .whenPrimitive(
                  classifyValue(
                      property.getKey(),
                      stringProperties,
                      numberProperties,
                      boolProperties,
                      pickProperties));
        }
        final Map<String, EditorValue> updates = shape.value;
        for (Map.Entry<String, EditorValue> updatedProperty : updates.entrySet()) {
          final SimpleEditorValue newValue =
              SimpleEditorValue.fromEditorValueOrNull(updatedProperty.getValue());
          final String propertyKey = updatedProperty.getKey();
          final SimpleEditorValue oldValue = properties.get(propertyKey);
          if (newValue != null && oldValue != null && newValue.type == oldValue.type) {
            newValue.value.whenPrimitive(
                classifyValue(
                    propertyKey,
                    stringProperties,
                    numberProperties,
                    boolProperties,
                    pickProperties));
          }
        }
        final T newValue =
            propertyEditor.writeProperties(
                value, stringProperties, numberProperties, boolProperties, pickProperties);
        EditorUtils.setNodeUNSAFE(f, node, newValue);
        return null;
      }
    };
  }

  private static EditorValue.EditorPrimitiveVisitor classifyValue(
      final String key,
      final Map<String, String> stringProperties,
      final Map<String, Number> numberProperties,
      final Map<String, Boolean> boolProperties,
      final Map<String, String> pickProperties) {
    return new EditorValue.EditorPrimitiveVisitor() {
      @Override
      public boolean isNumber(String[] path, EditorNumber number) {
        numberProperties.put(key, number.value);
        return false;
      }

      @Override
      public boolean isString(String[] path, EditorString string) {
        stringProperties.put(key, string.value);
        return false;
      }

      @Override
      public boolean isBool(String[] path, EditorBool bool) {
        boolProperties.put(key, bool.value);
        return false;
      }

      @Override
      public boolean isPick(String[] path, EditorPick pick) {
        pickProperties.put(key, pick.selected);
        return false;
      }
    };
  }

  // Types

  public interface PropertyReader<T> {
    Map<String, SimpleEditorValue> readProperties(T value);
  }

  public interface MutablePropertyEditor<T> extends PropertyReader<T> {
    void writeStringProperty(T value, String property, String newValue);

    void writeNumberProperty(T value, String property, Number newValue);

    void writeBoolProperty(T value, String property, boolean newValue);

    void writePickProperty(T value, String property, String newValue);
  }

  public abstract static class DefaultMutablePropertyEditor<T> implements MutablePropertyEditor<T> {

    @Override
    public void writeStringProperty(T value, String property, String newValue) {}

    @Override
    public void writeNumberProperty(T value, String property, Number newValue) {}

    @Override
    public void writeBoolProperty(T value, String property, boolean newValue) {}
  }

  public interface ImmutablePropertyEditor<T> extends PropertyReader<T> {
    T writeProperties(
        T value,
        Map<String, String> newStringValues,
        Map<String, Number> newNumberValues,
        Map<String, Boolean> newBoolValues,
        Map<String, String> newPickValues);
  }

  private static final int PRIMITIVE_TYPE_NUMBER = 0;
  private static final int PRIMITIVE_TYPE_STRING = 1;
  private static final int PRIMITIVE_TYPE_BOOL = 2;
  private static final int PRIMITIVE_TYPE_PICK = 3;

  /** A class that wraps either a number, string or bool */
  public static final class SimpleEditorValue {

    public final EditorValue value;

    public final int type;

    private SimpleEditorValue(EditorValue value, int type) {
      this.value = value;
      this.type = type;
    }

    // Constructors

    public static SimpleEditorValue number(Number n) {
      return new SimpleEditorValue(new EditorNumber(n), PRIMITIVE_TYPE_NUMBER);
    }

    public static SimpleEditorValue string(String s) {
      return new SimpleEditorValue(new EditorString(s), PRIMITIVE_TYPE_STRING);
    }

    public static SimpleEditorValue bool(boolean b) {
      return new SimpleEditorValue(new EditorBool(b), PRIMITIVE_TYPE_BOOL);
    }

    public static SimpleEditorValue pick(Set<String> otherValues, String selected) {
      return new SimpleEditorValue(new EditorPick(otherValues, selected), PRIMITIVE_TYPE_PICK);
    }

    public static @Nullable SimpleEditorValue fromEditorValueOrNull(EditorValue v) {
      return v.when(asPrimitive);
    }

    private static final EditorValue.EditorVisitor</*@Nullable*/ SimpleEditorValue> asPrimitive =
        new EditorValue.EditorVisitor</*@Nullable*/ SimpleEditorValue>() {
          @Override
          public @Nullable SimpleEditorValue isShape(EditorShape object) {
            return null;
          }

          @Override
          public @Nullable SimpleEditorValue isArray(EditorArray array) {
            return null;
          }

          @Override
          public @Nullable SimpleEditorValue isPick(EditorPick pick) {
            return SimpleEditorValue.pick(pick.values, pick.selected);
          }

          @Override
          public @Nullable SimpleEditorValue isNumber(EditorNumber number) {
            return SimpleEditorValue.number(number.value);
          }

          @Override
          public @Nullable SimpleEditorValue isString(EditorString string) {
            return SimpleEditorValue.string(string.value);
          }

          @Override
          public @Nullable SimpleEditorValue isBool(EditorBool bool) {
            return SimpleEditorValue.bool(bool.value);
          }
        };
  }
}

File: litho-editor-core/src/main/java/com/facebook/litho/editor/SimpleEditor.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho.editor

import com.facebook.litho.editor.Reflection.getValueUNSAFE
import com.facebook.litho.editor.Reflection.setValueUNSAFE
import com.facebook.litho.editor.model.EditorArray
import com.facebook.litho.editor.model.EditorBool
import com.facebook.litho.editor.model.EditorColor
import com.facebook.litho.editor.model.EditorNumber
import com.facebook.litho.editor.model.EditorPick
import com.facebook.litho.editor.model.EditorShape
import com.facebook.litho.editor.model.EditorString
import com.facebook.litho.editor.model.EditorValue
import com.facebook.litho.editor.model.EditorValue.DefaultEditorVisitor
import com.facebook.litho.editor.model.EditorValue.EditorPrimitiveVisitor
import java.lang.reflect.Field
import java.util.concurrent.atomic.AtomicReference

object SimpleEditor {
  // Constructors
  @JvmStatic
  fun <T> makeMutable(propertyEditor: MutablePropertyEditor<T>): Editor {
    return object : Editor {
      override fun read(f: Field, node: Any?): EditorValue {
        return makeValue<T>(f, node, propertyEditor)
      }

      override fun write(f: Field, node: Any?, values: EditorValue): Boolean {
        val value = getValueUNSAFE<T>(f, node) ?: return false
        values.`when`(writeMutable(propertyEditor, value))
        return true
      }
    }
  }

  @JvmStatic
  fun <T> makeImmutable(propertyEditor: ImmutablePropertyEditor<T?>): Editor {
    return object : Editor {
      override fun read(f: Field, node: Any?): EditorValue {
        return makeValue<T?>(f, node, propertyEditor)
      }

      override fun write(f: Field, node: Any?, values: EditorValue): Boolean {
        val value = getValueUNSAFE<T>(f, node)
        values.`when`(writeImmutable(f, node, propertyEditor, value))
        return true
      }
    }
  }

  // Read data
  private fun <T> makeValue(f: Field, node: Any?, propertyReader: PropertyReader<T>): EditorValue {
    val value = getValueUNSAFE<T>(f, node) ?: return EditorValue.shape(emptyMap())
    val properties = propertyReader.readProperties(value)
    val shape: MutableMap<String, EditorValue> = HashMap()
    for ((key, value1) in properties) {
      shape[key] = value1.value
    }
    return EditorValue.shape(shape)
  }

  // Write data
  private fun <T> writeMutable(
      propertyEditor: MutablePropertyEditor<T>,
      value: T
  ): DefaultEditorVisitor {
    return object : DefaultEditorVisitor() {
      override fun isShape(shape: EditorShape): Void? {
        val properties = propertyEditor.readProperties(value)
        val updates = shape.value
        for ((key, value1) in updates) {
          val newValue = SimpleEditorValue.fromEditorValueOrNull(value1)
          val oldValue = properties[key]
          if (newValue != null && oldValue != null) {
            val updatedValue = maybePickFromString(oldValue, newValue)
            if (updatedValue.type == oldValue.type) {
              updatedValue.value.whenPrimitive(
                  object : EditorPrimitiveVisitor {
                    override fun isNumber(path: Array<String>, number: EditorNumber): Boolean {
                      propertyEditor.writeNumberProperty(value, key, number.value)
                      return true
                    }

                    override fun isColor(path: Array<String>, color: EditorColor): Boolean {
                      propertyEditor.writeNumberProperty(value, key, color.value)
                      return true
                    }

                    override fun isString(path: Array<String>, string: EditorString): Boolean {
                      propertyEditor.writeStringProperty(value, key, string.value)
                      return false
                    }

                    override fun isBool(path: Array<String>, bool: EditorBool): Boolean {
                      propertyEditor.writeBoolProperty(value, key, bool.value)
                      return false
                    }

                    override fun isPick(path: Array<String>, pick: EditorPick): Boolean {
                      propertyEditor.writePickProperty(value, key, pick.selected)
                      return false
                    }
                  })
            }
          }
        }
        return null
      }
    }
  }

  private fun <T> writeImmutable(
      f: Field,
      node: Any?,
      propertyEditor: ImmutablePropertyEditor<T>,
      value: T
  ): EditorValue.EditorVisitor<Void> {
    return object : DefaultEditorVisitor() {
      override fun isShape(shape: EditorShape): Void? {
        val properties = propertyEditor.readProperties(value)
        val stringProperties: MutableMap<String, String> = HashMap()
        val numberProperties: MutableMap<String, Number> = HashMap()
        val boolProperties: MutableMap<String, Boolean> = HashMap()
        val pickProperties: MutableMap<String, String> = HashMap()
        for ((key, value1) in properties) {
          value1.value.whenPrimitive(
              classifyValue(
                  key, stringProperties, numberProperties, boolProperties, pickProperties))
        }
        val updates = shape.value
        for ((propertyKey, value1) in updates) {
          val newValue = SimpleEditorValue.fromEditorValueOrNull(value1)
          val oldValue = properties[propertyKey]
          if (newValue != null && oldValue != null) {
            val updatedValue = maybePickFromString(oldValue, newValue)
            if (updatedValue.type == oldValue.type) {
              updatedValue.value.whenPrimitive(
                  classifyValue(
                      propertyKey,
                      stringProperties,
                      numberProperties,
                      boolProperties,
                      pickProperties))
            }
          }
        }
        val newValue =
            propertyEditor.writeProperties(
                value, stringProperties, numberProperties, boolProperties, pickProperties)
        setValueUNSAFE(f, node, newValue)
        return null
      }
    }
  }

  // As Flipper doesn't have typed messages, selected Pick values are returned as String
  // and we have to apply the value to the old Pick
  private fun maybePickFromString(
      oldValue: SimpleEditorValue,
      newValue: SimpleEditorValue
  ): SimpleEditorValue {
    if (oldValue.type != PRIMITIVE_TYPE_PICK || newValue.type != PRIMITIVE_TYPE_STRING) {
      return newValue
    }
    val ref = AtomicReference(newValue)
    oldValue.value.`when`<Void>(
        object : DefaultEditorVisitor() {
          override fun isPick(pick: EditorPick): Void? {
            return newValue.value.`when`<Void>(
                object : DefaultEditorVisitor() {
                  override fun isString(string: EditorString): Void? {
                    ref.set(SimpleEditorValue.pick(pick.values, string.value))
                    return null
                  }
                })
          }
        })
    return ref.get()
  }

  private fun classifyValue(
      key: String,
      stringProperties: MutableMap<String, String>,
      numberProperties: MutableMap<String, Number>,
      boolProperties: MutableMap<String, Boolean>,
      pickProperties: MutableMap<String, String>
  ): EditorPrimitiveVisitor {
    return object : EditorPrimitiveVisitor {
      override fun isNumber(path: Array<String>, number: EditorNumber): Boolean {
        numberProperties[key] = number.value
        return false
      }

      override fun isColor(path: Array<String>, color: EditorColor): Boolean {
        numberProperties[key] = color.value
        return false
      }

      override fun isString(path: Array<String>, string: EditorString): Boolean {
        stringProperties[key] = string.value
        return false
      }

      override fun isBool(path: Array<String>, bool: EditorBool): Boolean {
        boolProperties[key] = bool.value
        return false
      }

      override fun isPick(path: Array<String>, pick: EditorPick): Boolean {
        pickProperties[key] = pick.selected
        return false
      }
    }
  }

  private const val PRIMITIVE_TYPE_NUMBER = 0
  private const val PRIMITIVE_TYPE_STRING = 1
  private const val PRIMITIVE_TYPE_BOOL = 2
  private const val PRIMITIVE_TYPE_PICK = 3

  // Types
  interface PropertyReader<T> {
    fun readProperties(value: T): Map<String, SimpleEditorValue>
  }

  interface MutablePropertyEditor<T> : PropertyReader<T> {
    fun writeStringProperty(value: T, property: String, newValue: String)

    fun writeNumberProperty(value: T, property: String, newValue: Number)

    fun writeBoolProperty(value: T, property: String, newValue: Boolean)

    fun writePickProperty(value: T, property: String, newValue: String)
  }

  abstract class DefaultMutablePropertyEditor<T> : MutablePropertyEditor<T> {
    override fun writeStringProperty(value: T, property: String, newValue: String) {}

    override fun writeNumberProperty(value: T, property: String, newValue: Number) {}

    override fun writeBoolProperty(value: T, property: String, newValue: Boolean) {}

    override fun writePickProperty(value: T, property: String, newValue: String) {}
  }

  interface ImmutablePropertyEditor<T> : PropertyReader<T> {
    fun writeProperties(
        value: T,
        newStringValues: Map<String, String>,
        newNumberValues: Map<String, Number>,
        newBoolValues: Map<String, Boolean>,
        newPickValues: Map<String, String>
    ): T
  }

  /** A class that wraps either a number, string, bool or pick */
  class SimpleEditorValue private constructor(val value: EditorValue, val type: Int) {
    companion object {
      // Constructors
      @JvmStatic
      fun number(n: Number): SimpleEditorValue {
        return SimpleEditorValue(EditorNumber(n), PRIMITIVE_TYPE_NUMBER)
      }

      @JvmStatic
      fun color(n: Number): SimpleEditorValue {
        return SimpleEditorValue(EditorNumber(n), PRIMITIVE_TYPE_NUMBER)
      }

      @JvmStatic
      fun string(s: String): SimpleEditorValue {
        return SimpleEditorValue(EditorString(s), PRIMITIVE_TYPE_STRING)
      }

      @JvmStatic
      fun bool(b: Boolean): SimpleEditorValue {
        return SimpleEditorValue(EditorBool(b), PRIMITIVE_TYPE_BOOL)
      }

      @JvmStatic
      fun pick(otherValues: Set<String>, selected: String): SimpleEditorValue {
        return SimpleEditorValue(EditorPick(otherValues, selected), PRIMITIVE_TYPE_PICK)
      }

      @JvmStatic
      fun fromEditorValueOrNull(v: EditorValue): SimpleEditorValue? {
        return v.`when`(asPrimitive)
      }

      private val asPrimitive: EditorValue.EditorVisitor<SimpleEditorValue> =
          object : EditorValue.EditorVisitor<SimpleEditorValue> {
            override fun isShape(`object`: EditorShape): SimpleEditorValue? {
              return null
            }

            override fun isArray(array: EditorArray): SimpleEditorValue? {
              return null
            }

            override fun isPick(pick: EditorPick): SimpleEditorValue? {
              return pick(pick.values, pick.selected)
            }

            override fun isNumber(number: EditorNumber): SimpleEditorValue? {
              return number(number.value)
            }

            override fun isColor(color: EditorColor): SimpleEditorValue? {
              return color(color.value)
            }

            override fun isString(string: EditorString): SimpleEditorValue? {
              return string(string.value)
            }

            override fun isBool(bool: EditorBool): SimpleEditorValue? {
              return bool(bool.value)
            }
          }
    }
  }
}
